tests/v2/validation/provisioning/rke2/etcd_bkp_restore_test.go

```diff
@@ -0,0 +1,504 @@
+	networkingv1 "k8s.io/api/networking/v1"
+	kwait "k8s.io/apimachinery/pkg/util/wait"
+)
+
+const (
+	defaultNamespace = "default"
+)
+
+type RKE2EtcdSnapshotRestoreTestSuite struct {
+	suite.Suite
+	session            *session.Session
+	client             *rancher.Client
+	clusterName        string
+	ns                 string
+	kubernetesVersions []string
+	cnis               []string
+	providers          []string
+	nodesAndRoles      []machinepools.NodeRoles
+}
+
+func (p *RKE2EtcdSnapshotRestoreTestSuite) TearDownSuite() {
+	p.session.Cleanup()
+}
+
+var EtcdSnapshotGroupVersionResource = schema.GroupVersionResource{
+	Group:    "rke.cattle.io",
+	Version:  "v1",
+	Resource: "etcdsnapshots",
+}
+
+func (r *RKE2EtcdSnapshotRestoreTestSuite) SetupSuite() {
+	testSession := session.NewSession(r.T())
+	r.session = testSession
+
+	r.ns = defaultNamespace
+
+	clustersConfig := new(provisioning.Config)
+	config.LoadConfig(provisioning.ConfigurationFileKey, clustersConfig)
+
+	r.kubernetesVersions = clustersConfig.KubernetesVersions
+	r.cnis = clustersConfig.CNIs
+	r.providers = clustersConfig.Providers
+	r.nodesAndRoles = clustersConfig.NodesAndRoles
+
+	client, err := rancher.NewClient("", testSession)
+	require.NoError(r.T(), err)
+
+	r.client = client
+
+	r.clusterName = r.client.RancherConfig.ClusterName
+}
+
+func (r *RKE2EtcdSnapshotRestoreTestSuite) createSnapshot(client *rancher.Client, clustername string, generation int) error {
+	clusterObj, existingSteveAPIObj, err := clusters.GetProvisioningClusterByName(client, clustername, r.ns)
+	if err != nil {
+		return err
+	}
+
+	clusterObj.Spec.RKEConfig.ETCDSnapshotCreate = &rkev1.ETCDSnapshotCreate{
+		Generation: generation,
+	}
+
+	_, err = client.Steve.SteveType(clusters.ProvisioningSteveResouceType).Update(existingSteveAPIObj, clusterObj)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *RKE2EtcdSnapshotRestoreTestSuite) restoreSnapshot(client *rancher.Client, clustername string, name string,
+	generation int, restoreconfig string) error {
+
+	clusterObj, existingSteveAPIObj, err := clusters.GetProvisioningClusterByName(client, clustername, r.ns)
+	if err != nil {
+		return err
+	}
+
+	clusterObj.Spec.RKEConfig.ETCDSnapshotRestore = &rkev1.ETCDSnapshotRestore{
+		Name:             name,
+		Generation:       generation,
+		RestoreRKEConfig: restoreconfig,
+	}
+
+	_, err = client.Steve.SteveType(clusters.ProvisioningSteveResouceType).Update(existingSteveAPIObj, clusterObj)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *RKE2EtcdSnapshotRestoreTestSuite) GetSnapshots(client *rancher.Client,
+	localClusterID string) ([]provisioningV1.SteveAPIObject, error) {
+
+	steveclient, err := client.Steve.ProxyDownstream(localClusterID)
+	if err != nil {
+		return nil, err
+	}
+	snapshotSteveObjList, err := steveclient.SteveType("rke.cattle.io.etcdsnapshot").List(&types.ListOpts{})
+	if err != nil {
+		return nil, err
+	}
+
+	return snapshotSteveObjList.Data, nil
+
+}
```

Comment:
I completely agree with this. If there are assertions, or specifics from r then yes they belong here. But since they don't this is either an extension or a helper in companion` yourtest.go` file

Reply:
Created the separate etcd_bkp_restore.go file and moved all the related things.